Broker Reconciliation occurs too frequently

jboyd01 opened the pull request:
fixes #1513 

Added a new field for Broker's last reconciled time.  When determining if we should reconcile the broker, check that the last reconciled time vs the Ready condition's LastTransitionTime as LTT won't change over time under normal good circumstances when status remains Ready=True.

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -32,8 +32,9 @@ type ClusterServiceBroker struct {
-	Status ClusterServiceBrokerStatus
+	Spec           ClusterServiceBrokerSpec
+	Status         ClusterServiceBrokerStatus
+	LastReconciled metav1.Time
```
This field should be part of status and should be called LastCatalogRetrievalTime, and should be a pointer (zero value time is not meaningful)

staebler commented:
Would it make sense to update the Ready condition every time that the Controller successfully fetches the catalog from the broker? Logically, we are rebuilding the broker, so it doesn't seem inappropriate to me to change the transition time on the Ready condition.

jboyd01 commented:
@staebler  we considered updating the LastTransitionTime but that felt like it violated the semantics

jboyd01 commented:
@pmorie I've updated per your review

pmorie commented on pkg/controller/controller_broker.go:
```diff
@@ -128,11 +128,14 @@ func shouldReconcileClusterServiceBroker(broker *v1beta1.ClusterServiceBroker, n
+				}
 				if intervalPassed == false {
-					glog.V(10).Info(pcb.Message("Not processing because RelistDuration has not elapsed since the broker became ready"))
+					glog.V(10).Info(pcb.Message("Not processing because RelistDuration has not elapsed since the last relist."))
```
nit: no period

jboyd01 replied:
ok, addressed

kibbles-n-bytes commented:
@jboyd01 Why do you feel it violated the semantics? I aggree with @staebler; this is a new event that is being logged with a new `LastTransitionTime`. It just happens to be that the result of that event is the same as the previous successful fetching attempt.

jboyd01 commented:
@kibbles-n-bytes I'm viewing LastTransitionTime as transition from one state to another, ie Ready -> Not Ready...  I'd could be convinced otherwise, but in chatting with @pmorie he was in agreement.  Paul?

staebler commented:
@jboyd01 You're right. I was using the scenario where the controller sets a Ready/False condition for a resource that already has a Ready/False condition. I thought that in that scenario the transition time was updated to the current time. I see now that the transition time is not changed.

kibbles-n-bytes commented:
@jboyd01 Ah, I see. I thought the same as @staebler.

kibbles-n-bytes commented:
@jboyd01 Can we get a test for the state where the relist duration has elapsed with respect to the `LastTransitionTime`, but not the `LastCatalogRetrievalTime`, and so the catalog is not relisted?

jboyd01 commented:
@kibbles-n-bytes Absolutely, updated tests should have been included.  I've updated adding an explicit test for:
 1.  LastTransitionTime a long time ago & LastCatalogRetrievalTime meets duration expiration (results in **relist**)
 2.  LastTransitionTime a long time ago & LastCatalogRetrievalTime fails to meet duration expiration (it was just relisted, **no relist**)
